File: algorithms/393. UTF-8 Validation/solution.cc
class Solution {
public:
    bool validUtf8(vector<int>& data) {
        int follow = 0;
        for ( int n : data ) {
            if ( follow ) {
                if ( n >> 6 != 0b10 ) {
                    return false;
                }
                --follow;
            } else {
                if ( n >> 5 == 0b110 ) {
                    follow = 1;
                } else if ( n >> 4 == 0b1110 ) {
                    follow = 2;
                } else if ( n >> 3 == 0b11110 ) {
                    follow = 3;
                } else if ( n >> 7 ) {
                    return false;
                }
            }
        }
        return follow == 0;
    }
};

File: algorithms/393. UTF-8 Validation/solution.py
class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        follow = 0
        for n in data:
            if follow != 0:
                if n >> 6 != 0b10:
                    return False
                follow -= 1
            elif n >> 5 == 0b110:
                follow = 1
            elif n >> 4 == 0b1110:
                follow = 2
            elif n >> 3 == 0b11110:
                follow = 3
            elif n >> 7:
                return False
        return follow == 0
